### PR TITLE
ARTEMIS-1221 Duplicated ID causes LargeMessage lost at backup

### PR DESCRIPTION
**Pull request by gaohoward**
When a large message is replicated to backup, a pendingID is generated
when the large message is finished. This pendingID is generated by a
BatchingIDGenerator at backup.

It is possible that a pendingID generated at backup may be a duplicate
to an ID generated at live server.

This can cause a problem when a large message with a messageID that is
the same as another largemessage's pendingID is replicated and stored
in the backup's journal, and then a deleteRecord for the pendingID
is appended. If backup becomes live and loads the journal, it will
drop the large message add record because there is a deleteRecord of
the same ID (even though it is a pendingID of another message).
As a result the expecting client will never get this large message.

So in summary, the root cause is that the pendingIDs for large
messages are generated at backup while backup is not alive.
The solution to this is that instead of the backup generating
the pendingID, we make them all be generated in advance
at live server and let them replicated to backup whereever needed.
The ID generater at backup only works when backup becomes live
(when it is properly initialized from journal).

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/ReplicationLargeMessageBeginMessage.java [hunk 1]
```diff
@@ -44,11 +46,13 @@ public int expectedEncodeSize() {
    @Override
    public void encodeRest(final ActiveMQBuffer buffer) {
       buffer.writeLong(messageId);
```
Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/protocol/core/impl/wireformat/ReplicationLargeMessageBeginMessage.java]: I'm not sure how this is possibly working...


expectedEncodeSize is not taking this new long into consideration...


Besides this will cause versioning issues...


So, you will need to do something like this on reading... 

On reading.. you will need to verify the positioning of the buffer.. against the declared size.. if there's another long coming.. otherwise it wouldn't read it.. and use the current semantic.


this is a -1.. can you fix it?

**Comment by gaohoward** [comment]
@clebertsuconic ok, I'll fix it.

**Comment by gaohoward** [comment]
@clebertsuconic 
well, it turns out more work need to be done. It happens that the changes of code just avoids the bug and makes the test pass. The pendingIDs are not used at all. I need a bit more time to correct it. :)

**Comment by gaohoward** [comment]
@clebertsuconic I think I've done it correctly this time. Pls review again.
Thanks
Howard

**Comment by clebertsuconic** [comment]
whoa.. that's a nice Fix!!! Very nice!!!!

**Comment by gaohoward** [comment]
@clebertsuconic Thanks! :)
